src/layouts/app/administrative/users/edit/index.tsx

```diff
@@ -1,6 +1,180 @@
+    []
+  );
+  if (!isControllerSubscribed && !initialized) {
+    return <h1>Preloader</h1>;
+  }
+
+  return <Formik
+    enableReinitialize
+    onSubmit={onSubmit}
+    initialValues={initialValues}
+  >
+    <Form>
+      <Grid container spacing={2}>
+        <Grid item xs={12}>
+          <Typography variant="h3">{ id === NEW_ID ? 'Create' : 'Edit' } User</Typography>
+        </Grid>
+        <Grid item container direction="column" xs={6}>
+          <Grid item mb={2}>
+            <Card>
+              <CardHeader title={<Typography variant="h4">Main</Typography>} />
+              <Divider />
+              <CardContent>
+                <Box>
+                  <FInput
+                    required
+                    type="text"
+                    size="small"
+                    name="firstName"
+                    label={<strong>First Name</strong>}
+                  />
+                </Box>
+                <Box pt={2}>
+                  <FInput
+                    type="text"
+                    size="small"
+                    name="middleName"
+                    label={<strong>Middle Name</strong>}
+                  />
+                </Box>
+                <Box pt={2}>
+                  <FInput
+                    required
+                    type="text"
+                    size="small"
+                    name="lastName"
+                    label={<strong>Last Name</strong>}
+                  />
+                </Box>
+                <Box pt={2}>
+                  <FSelect
+                    fullWidth
+                    required
+                    size="small"
+                    name="suffix"
+                    label={<strong>Suffix</strong>}
+                    getFieldValue={(({ value }: AnyObject) => value)}
+                    prepareValue={(value: string | number) => ({ value, label: value })}
+                    options={['Jr.', 'Sr.', '2nd', '3rd', 'II', 'III', 'IV', 'V'].map(option => ({ value: option, label: option }))}
+                  />
+                </Box>
+              </CardContent>
+            </Card>
+          </Grid>
+          <Grid item>
+            <Card>
+              <CardHeader title={<Typography variant="h4">Role</Typography>} />
+              <Divider />
+              <CardContent>
+                <Box>
+                  <FAsyncSelect
+                    multiple
+                    required
+                    fullWidth
+                    size="small"
+                    name="roles"
+                    loadingText="LOADING"
+                    filterSelectedOptions
+                    loadOptions={getRolesMemo}
+                    label={<strong>Roles</strong>}
+                    getOptionLabel={({ name }) => name}
+                    prepareValue={(value: AnyObject) => value}
+                    getFieldValue={(value) => value}
+                    isOptionEqualToValue={(option, value) => option.name === value.name}
```

Comment:
why do all functions aren't memoized? I suggest to extact them to constants/extractors.ts